AFQ/viz/utils.py

```diff
@@ -411,6 +434,615 @@ def visualize_tract_profiles(tract_profiles, scalar="dti_fa", min_fa=0.0,
+                for scalar, threshold in constraint.items():
+                    profile[scalar] = \
+                        profile[scalar].apply(
+                            lambda x: self._threshold_scalar(
+                                bound,
+                                threshold,
+                                x))
+
+            self.profile_dict[names[i]] = profile
+        if subjects is None:
+            self.subjects = self.profile_dict[names[0]]['subjectID'].unique()
+        else:
+            self.subjects = subjects
+
+    def _threshold_scalar(self, bound, threshold, val):
+        if bound == "lb":
+            if val > threshold:
+                return val
+            else:
+                return np.nan
+        elif bound == "ub":
+            if val < threshold:
+                return val
+            else:
+                return np.nan
+        else:
+            raise RuntimeError("scalar_bounds dictionary "
+                               + " formatted incorrectly. See"
+                               + " the default for reference")
+
+    def _warn_not_found(self, scalar, subject, bundle, name):
+        self.logger.warning(
+            'No scalars found for scalar ' + scalar
+            + ' for subject ' + str(subject)
+            + ' for bundle ' + bundle
+            + ' for CSV ' + name)
+
+    def _warn_nans(self, scalar, subject, bundle, name, repl_nan):
+        message = (
+            'NaNs found in scalar ' + scalar
+            + ' for subject ' + str(subject)
+            + ' for bundle ' + bundle
+            + ' for CSV ' + name)
+        if repl_nan:
+            message = message + '. These Nans were replaced with 0.'
+        self.logger.info(message)
+
+    def _warn_many_nans(self, scalar, subject, bundle, name):
+        self.logger.warning(
+            'More than 10 NaNs found in scalar ' + scalar
+            + ' for subject ' + str(subject)
+            + ' for bundle ' + bundle
+            + ' for CSV ' + name)
+
+    def _get_fname(self, folder, f_name):
+        f_folder = op.join(
+            self.out_folder,
+            folder)
+        os.makedirs(f_folder, exist_ok=True)
+        return op.join(f_folder, f_name)
+
+    def _get_profile(self, name, bundle, subject, scalar, repl_nan=True):
+        profile = self.profile_dict[name]
+        single_profile = profile[
+            (profile['subjectID'] == subject)
+            & (profile['bundle'] == bundle)
+        ][scalar].to_numpy()
+        nans = np.isnan(single_profile)
+        if len(single_profile) < 1:
+            self._warn_not_found(scalar, subject, bundle, name)
+            return None
+
+        if np.sum(nans) > 10:
+            self._warn_many_nans(scalar, subject, bundle, name)
+            return None
+
+        if np.sum(nans) > 0:
+            self._warn_nans(scalar, subject, bundle, name, repl_nan)
+            if repl_nan:
+                single_profile[nans] = 0
```

Comment:
IIUC, this means that we impute to 0. Another (maybe more careful?) approach would be to interpolate from the closest nodes, if this is in the middle of a profile, or extrapolate the last value out, if this is at the end of a profile.

Reply:
We can't use nan-agnostic functions like nanmean or nanstd here, right? Because you need a non-nan value to calculate correlations.

Reply:
If the profiles were only NaN's, they would not have made it this far. NaN agnostic functions calculate the mean using non-NaN values (ignore the NaNs) while the normal numpy functions simply propogates the NaNs.